File: tsp.cpp
#include <algorithm>
#include <cstddef>
#include <cstdint>
#include <cstdlib>
#include <fstream>
#include <iostream>
#include <random>

#include "genetic_algorithms.hpp"

struct town
{
    town(double x, double y) : x(x), y(y) {}

    bool operator==(const town& other) { return x == other.x && y == other.y; }

    double x;
    double y;
};

double fitness(const std::vector<int64_t> genome, const std::vector<town>& towns)
{
    double total_distance = 0;
    double x;
    double y;
    for (size_t i = 0; i < genome.size() - 1; ++i)
    {
        x = towns[genome[i]].x - towns[genome[i + 1]].x;
        y = towns[genome[i]].y - towns[genome[i + 1]].y;

        total_distance += std::sqrt(std::pow(x, 2) + std::pow(y, 2));
    }

    return total_distance;
}

int main(int argc, const char** argv)
{
    if (argc != 5)
    {
        std::cerr << "USAGE: ./tsp.x <n-towns> <population-size> <generations> <mutation-rate>"
                  << std::endl;
        return 1;
    }

    // generate random items
    std::random_device rd;
    std::mt19937 engine(rd());
    std::uniform_real_distribution<double> dist(50, 1000);

    size_t num_of_towns = std::stoul(argv[1]);
    std::vector<town> towns;
    std::vector<int64_t> genome_values;
    for (size_t i = 0; i < num_of_towns; ++i)
    {
        town new_town(dist(engine), dist(engine));

        while (std::find(towns.begin(), towns.end(), new_town) != towns.end())
            new_town = {dist(engine), dist(engine)};
        towns.push_back({dist(engine), dist(engine)});
        genome_values.push_back(i);
    }

    std::cout << "************ GENETIC ****************" << std::endl;
    size_t population_size = std::stoul(argv[2]);
    size_t generations = std::stoul(argv[3]);
    double mutation_rate = std::stod(argv[4]);

    ga<int64_t, double> genetic_tsp(population_size, num_of_towns, genome_values, generations,
                                    problem_type::minimization);
    genetic_tsp.shuffle_generate_population();
    genetic_tsp.evaluate_population(fitness, towns);

    auto best_individual = genetic_tsp.get_best_individual();
    std::cout << "best distance: " << best_individual.get_fitness() << std::endl;

    std::ofstream file("towns.csv");
    file << "x,y" << std::endl;
    for (size_t i = 0; i < num_of_towns; ++i)
        file << towns[i].x << "," << towns[i].y << std::endl;

    individual<int64_t, double> current_best;
    std::bernoulli_distribution mutation_dist(0.5);
    for (size_t g = 0; g < generations; ++g)
    {
        genetic_tsp.roulette();

        genetic_tsp.one_point_crossover_v2();

        if (mutation_dist(engine))
            genetic_tsp.rotate_mutation(mutation_rate);
        else
            genetic_tsp.swap_mutation(mutation_rate);

        genetic_tsp.evaluate_offsprings(fitness, towns);

        genetic_tsp.replace();

        genetic_tsp.evaluate_population(fitness, towns);

        current_best = genetic_tsp.get_best_individual();
        if (current_best > best_individual)
            best_individual = current_best;
    }

    best_individual = genetic_tsp.get_best_individual();
    std::cout << "best distance: " << best_individual.get_fitness() << std::endl;
    std::ofstream solution("solution.txt");
    for (size_t i = 0; i < num_of_towns; ++i)
        solution << best_individual[i] << std::endl;

    return 0;
}

File: scripts/tsp.py
import math
import random
import sys

import genetic
import matplotlib.pyplot as plt
import pandas as pd


class town:
    def __init__(self, x: float, y: float):
        self.x = x
        self.y = y


def fitness(genome: list[int], towns: list[town]) -> float:
    total_distance = 0
    for i in range(len(towns) - 1):
        x = towns[genome[i]].x - towns[genome[i + 1]].x
        y = towns[genome[i]].y - towns[genome[i + 1]].y

        total_distance += math.sqrt(math.pow(x, 2) + math.pow(y, 2))

    return total_distance / (len(towns) - 1)


if __name__ == "__main__":
    if len(sys.argv) != 5:
        print(
            "USAGE: py tsp.py <n-towns> <population-size> <generations> <mutation-rate>"
        )
        exit(-1)

    num_of_towns = int(sys.argv[1])
    towns: list[town] = []
    genome_values = []
    for i in range(num_of_towns):
        t = town(random.uniform(50, 100), random.uniform(50, 100))
        while t in towns:
            t = town(random.uniform(50, 100), random.uniform(50, 100))
        towns.append(t)
        genome_values.append(i)

    population_size = int(sys.argv[2])
    generations = int(sys.argv[3])
    mutation_rate = float(sys.argv[4])

    tsp = genetic.ga(
        population_size,
        num_of_towns,
        genome_values,
        generations,
        genetic.problem_type.minimization,
    )

    tsp.shuffle_generate_population()
    tsp.evaluate_population(fitness, towns)

    best_individual = tsp.get_best_individual()
    print(f"best distance: {best_individual.get_fitness()}")

    for g in range(generations):
        tsp.roulette()
        tsp.one_point_crossover_v2()
        tsp.rotate_mutation(mutation_rate)
        tsp.evaluate_offsprings(fitness, towns)
        tsp.replace()
        tsp.evaluate_population(fitness, towns)

    best_individual = tsp.get_best_individual()
    print(f"best distance: {best_individual.get_fitness()}")


    for t in towns:
        x = t.x
        y = t.y

    file = open("solution.txt", "r")
    line = file.readline()
    solution = []
    while line != "":
        solution.append(int(line))
        line = file.readline()

    x_new = [x[solution[i]] for i in range(len(x))]
    y_new = [y[solution[i]] for i in range(len(y))]

    plt.figure(figsize=(10, 6))

    plt.scatter(x, y, c="r", label="middle town")
    plt.scatter([x_new[0]], [y_new[0]], c="g", label="start town")
    plt.scatter([x_new[-1]], [y_new[-1]], c="b", label="final town")
    plt.plot(x_new, y_new, c="k", label="path")

    plt.title("Best path found")
    plt.xlabel("X town coordinate")
    plt.ylabel("Y town coordinate")
    plt.legend()
    plt.show()
